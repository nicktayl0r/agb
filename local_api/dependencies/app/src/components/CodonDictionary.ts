export const CODON_DICTIONARY: { [index: string]: string } = {
  UUU: "Phenylalanine",
  UUC: "Phenylalanine",
  UUA: "Leucine",
  UUG: "Leucine",
  CUU: "Leucine",
  CUC: "Leucine",
  CUA: "Leucine",
  CUG: "Leucine",
  AUU: "Isoleucine",
  AUC: "Isoleucine",
  AUA: "Isoleucine",
  AUG: "Methionine",
  GUU: "Valine",
  GUC: "Valine",
  GUA: "Valine",
  GUG: "Valine",
  UCU: "Serine",
  UCC: "Serine",
  UCA: "Serine",
  UCG: "Serine",
  CCU: "Proline",
  CCC: "Proline",
  CCA: "Proline",
  CCG: "Proline",
  ACU: "Threonine",
  ACC: "Threonine",
  ACA: "Threonine",
  ACG: "Threonine",
  GCU: "Alanine",
  GCC: "Alanine",
  GCA: "Alanine",
  GCG: "Alanine",
  UAU: "Tyrosine",
  UAC: "Tyrosine",
  UAA: "Stop",
  UAG: "Stop",
  CAU: "Histidine",
  CAC: "Histidine",
  CAA: "Glutamine",
  CAG: "Glutamine",
  AAU: "Asparagine",
  AAC: "Asparagine",
  AAA: "Lysine",
  AAG: "Lysine",
  GAU: "Aspartic acid",
  GAC: "Aspartic acid",
  GAA: "Glutamic acid",
  GAG: "Glutamic acid",
  UGU: "Cysteine",
  UGC: "Cysteine",
  UGA: "Stop_UGA",
  UGG: "Tryptophan",
  CGU: "Arginine",
  CGC: "Arginine",
  CGA: "Arginine",
  CGG: "Arginine",
  AGU: "Serine",
  AGC: "Serine",
  AGA: "Arginine",
  AGG: "Arginine",
  GGU: "Glycine",
  GGC: "Glycine",
  GGA: "Glycine",
  GGG: "Glycine"
};
