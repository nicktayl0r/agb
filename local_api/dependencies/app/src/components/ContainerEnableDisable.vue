<template>
  <fieldset :id="id" v-show="visibleStore" :disabled="disabled" :class="[ 'no-border', AddRemoveClasses]" ><slot/></fieldset>
</template>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";
import WidgetEnabled from "../mixins/WidgetEnabled";

@Component({ mixins: [WidgetEnabled] })
export default class ContainerEnableDisable extends Widget {}
</script>
