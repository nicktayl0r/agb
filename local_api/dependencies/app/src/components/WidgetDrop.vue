// I chose fieldset over span because it has a disabled attribute
<template>
	<fieldset
    :id="id"
    v-show="visibleStore"
    :disabled="disabled"
    @drop="drop"
    @dragover="dragover"
    @dragenter="dragenter"
    class="drop"
    :class="AddRemoveClasses"
    :maxChildren="maxChildrenStore"
    :overflowStyle="overflowStyleStore">
		<span class="contentSlot noSelect">
			<slot />
		</span>
		<span class="dragSlot"></span>
	</fieldset>
</template>

<script lang="ts">
import { Maybe } from "true-myth";

import { Component, Vue, Prop } from "vue-property-decorator";
import Widget from "./Widget";
import { logMessage, logWarning } from "@/helpers/debugHelpers";
import WidgetEnabled from "@/mixins/WidgetEnabled";
import { readWidgetPropVal } from "@/store/modules/appData";
import {
  readWidgetEntryVal,
  dispatchUpdateWidget
} from "@/store/modules/userData";
import { widgetValPrimitive } from "../data models/widgetModels";
import { getSharedDataUserOrAppVal } from "@/helpers/sharedDataHelpers";
import { getStore } from "@/helpers/storeHelpers";

@Component({ mixins: [WidgetEnabled] })
export default class WidgetDrop extends Widget {
	@Prop({ default: 1 })
	maxChildren: number;

	@Prop({ default: "reset-child" })
	overflowStyle: "reset-child" | "swap";

  @Prop({ default: "" })
	allowedGroupId: string;

	// written into the store in WidgetDrag, because drags can be moved by effects
	get maxChildrenStore() {
	  const max = this.propStoreOverride("maxChildren") as number;
	  if (this.$el) {
	    const dragSlot = this.$el.getElementsByClassName("dragSlot")[0];
	    if (dragSlot.children.length > max) {
	      const extra = dragSlot.children.length - max;
	      for (let i = 0; i < extra; i++) {
	        const extraChild = dragSlot.children[
	          dragSlot.children.length - 1 - i
	        ] as HTMLElement;
	        const extraChildID = extraChild.id;
	        const extraChildWriteKey = extraChild.getAttribute(
	          "shared-data-write-key"
	        );
	        const startDropID = this.GetPropDropID(extraChildID);
          this.UpdateChildDropID(
	          extraChildID,
	          extraChildWriteKey,
	          extraChild,
	          startDropID
	        );
	      }
	    }
	  }
	  return max;
	}

	get overflowStyleStore() {
	  return this.propStoreOverride("overflowStyle") as string;
	}

	mounted() {
	  // prevent child images in contentSlot from doing their default drag behavior
	  const contentSlot = this.$el.querySelector(".contentSlot");
	  if (contentSlot) {
	    const childIMGs = contentSlot.getElementsByTagName("IMG");
	    for (let i = 0; i < childIMGs.length; i++) {
	      childIMGs[i].setAttribute("draggable", "false");
	    }
	  }
	}

	dragover(ev: DragEvent) {
	  ev.preventDefault();
	}

	// needed by mobile polyfill
	dragenter(ev: DragEvent) {
	  ev.preventDefault();
	}

	drop(ev: DragEvent) {
	  ev.preventDefault();

	  if (!this.propStoreOverride("enabled") || !this.notGrapes) return;

	  const target = this.$el.getElementsByClassName("dragSlot")[0];

	  if (!ev.dataTransfer) {
	    logWarning("null ev.dataTransfer", ev);
	    return;
	  }
	  const draggedWidgetID = ev.dataTransfer.getData("draggedWidgetID");
	  const dragPropDropId: widgetValPrimitive = this.GetPropDropID(
	    draggedWidgetID
	  );
	  const dragStorePropId: widgetValPrimitive = this.GetStoreDropID(
	    draggedWidgetID
	  );

    const maybeDragGroupId = readWidgetPropVal(this.$store)(
      //@ts-ignore
	    this.pageID,
	    draggedWidgetID,
	    "groupId"
	  );

    const dragGroupId = maybeDragGroupId.isJust() ? maybeDragGroupId.unsafelyUnwrap() : ""

	  const droppedEl = document.getElementById(draggedWidgetID);
	  const writeKey = ev.dataTransfer.getData("draggedWidgetSharedDataWriteKey");
    if(this.allowedGroupId === "" || this.allowedGroupId === dragGroupId) {

      //go ahead and append the new drag widget. It's important in order to free up the other dropWidget.
      if (dragStorePropId == "") {
        //the drag's dropID is at its original value. Don't fire an update if the original dropID equals our dropID
        if (dragPropDropId !== this.id) {
          const dragSlot = this.$el.getElementsByClassName("dragSlot")[0];
            this.UpdateChildDropID(draggedWidgetID, writeKey, droppedEl, this.id);
        }
      } else if (dragStorePropId !== this.id) {
        // the drag being dropped isn't already in here
        const dragSlot = this.$el.getElementsByClassName("dragSlot")[0];
          this.UpdateChildDropID(draggedWidgetID, writeKey, droppedEl, this.id);
      }
      if (target.children.length >= this.maxChildrenStore) {
        //we are at the max children allowed, swap out the last drag
        //first send the last drag back to its starting drop
        const lastChildEL = target.children[target.children.length - 1];
        const lastChildID = lastChildEL.id;
        if (lastChildID == draggedWidgetID) return; //fixes jest tests. In jest - children.length is already 1, while in real dom children.length is 0.
        const lastChildPropId: widgetValPrimitive = this.GetPropDropID(
          lastChildID
        );
        const lastChildWriteKey: string = Maybe.unwrapOr(
          "",
          readWidgetPropVal(this.$store)(this.pageID, draggedWidgetID, "dropId")
        ).toString();

        const pageID = this.pageID;
        const store = this.$store;
        const prevDropAvailable = function(): boolean {
          const dropChildren = Array.from(document.querySelectorAll(`#${lastChildPropId} fieldset`));

          const dropMaxChildrenMaybe: any =  readWidgetPropVal(store)(
            pageID,
            lastChildPropId,
            "maxChildren"
          );
          const dropMaxChildren = dropMaxChildrenMaybe.unsafelyUnwrap();

          return dropChildren.length >= dropMaxChildren
            ? false
            : true
        }

        switch (this.overflowStyleStore) {
        case "reset-child":
          if(prevDropAvailable()) {
            this.UpdateChildDropID(
              lastChildID,
              lastChildWriteKey,
              lastChildEL as HTMLElement,
              lastChildPropId
            );
          }
          break;
        case "swap":
            if (dragStorePropId === "") {
              //send the lastChild to the dropID of the incoming
              if (dragPropDropId != this.id) {
                this.UpdateChildDropID(
                  lastChildID,
                  lastChildWriteKey,
                  lastChildEL as HTMLElement,
                  dragPropDropId
                );
              }
            } else {
              // the drag being dropped isn't already in here
              this.UpdateChildDropID(
                lastChildID,
                lastChildWriteKey,
                  lastChildEL as HTMLElement,
                  dragStorePropId
              );
              this.UpdateChildDropID(draggedWidgetID, writeKey, droppedEl, this.id);
            }
            break;
        }
      }
    }
	}

	UpdateChildDropID(
	  childID: string,
	  writeKey: string | null,
	  droppedEl: HTMLElement | null,
	  dropID: string
	) {
	  if (dropID) {
	    this.updateWidgetID(childID, "dropId", dropID);
	    if (writeKey) {
	      this.updateSharedDataVal(writeKey, dropID);
	    }
	    if (droppedEl) {
	      const droppedVm: Vue = (droppedEl as any).__vue__;
	      if (droppedVm) droppedVm.$forceUpdate();
	    }
	  }
	}

	GetPropDropID(draggedWidgetID: string) {
	  //get the dropID currently in the store
	  const maybePropVal = readWidgetPropVal(this.$store)(
	    this.pageID,
	    draggedWidgetID,
	    "dropId"
	  );
	  return Maybe.unwrapOr("", maybePropVal).toString();
	}

	GetStoreDropID(draggedWidgetID: string) {
	  const maybeStoreVal = readWidgetEntryVal(this.$store)(
	    this.pageID,
	    draggedWidgetID,
	    "dropId"
	  );
	  return Maybe.unwrapOr("", maybeStoreVal).toString();
	}
}

function getDragIDs(dropID: string, pageID: string) {
  //get value from store, convert to string[], retain string value if string
  const maybeDragIds = readWidgetEntryVal(getStore())(pageID, dropID, "dragId");
  let dragIdArray: string[] | undefined = undefined;
  if (maybeDragIds.isJust()) {
    const maybeDragIds_val = Maybe.unsafelyUnwrap(maybeDragIds);
    if (typeof maybeDragIds_val === "string" && maybeDragIds_val != "") {
      dragIdArray = [];
      dragIdArray.push(maybeDragIds_val);
    } else dragIdArray = (maybeDragIds_val as string[]).slice(); //important to copy the array. Don't edit store values directly.
  }
  return dragIdArray;
}

export function removeDragWidget(
  dropID: string,
  dragID: string,
  pageID: string
) {
  const dragIdArray = getDragIDs(dropID, pageID);
  if (dragIdArray == undefined) return;
  if (dragIdArray.includes(dragID)) {
    //remove from the array
    const removeIndex = dragIdArray.indexOf(dragID);
    dragIdArray.splice(removeIndex, 1);
  } else return;
  dispatchUpdateWidget(getStore(), {
    pageID,
    widgetID: dropID,
    key: "dragId",
    //@ts-ignore dragIdArray won't be undefined.
    value: dragIdArray
  });
}

export function addDragWidget(dropID: string, dragID: string, pageID: string) {
  let dragIdArray = getDragIDs(dropID, pageID);
  if (dragIdArray == undefined) dragIdArray = [];
  if (!dragIdArray.includes(dragID)) {
    dragIdArray.push(dragID);
  } else return;
  dispatchUpdateWidget(getStore(), {
    pageID,
    widgetID: dropID,
    key: "dragId",
    value: dragIdArray
  });
}
</script>

<style>
.drop {
	display: inline-block;
}
.buttonDrop {
  transition-property: transform;
	transition-duration: 0.4s;
	background-color: var(--color-white);
	/* opacity: 0%; */
	border: 0.4vh solid var(--color-grey2);
	border-radius: 1.3vh;
	min-width: 4.5vh;
	min-height: 4.5vh;
	line-height: 3vh;
	padding: 0;
	margin: 0 auto;
	text-align: center;
	vertical-align: middle;
}
.buttonDrop .contentSlot {
	/* this is to vertical align slot contents */
	display: flex;
	justify-content: center;
	flex-direction: column;
	height: 100%;
	z-index: 0;
}
.buttonDrop .dragSlot {
	/* this is to vertical align slot contents */
	display: flex;
	justify-content: center;
	flex-direction: column;
	min-width: 4.5vh;
	min-height: 4.5vh;
	line-height: 3vh;
	/* top: -10.25vh; */
	position: relative;
	z-index: 1;
}
.buttonDrop:enabled:hover {
	background-color: var(--color-white);
	border: 3px solid var(--color-blue3);
	border-radius: 5px;
  transition-property: transform;
	transition-duration: 0.4s;
}
.dropFlexHorizontal .dragSlot {
	flex-direction: row;
}
.dropFlexVertical .dragSlot {
	flex-direction: column;
}
.drop.dropDiamond {
	background: url("../assets/svg/dropDiamond.svg") no-repeat center;
	border: 0vh solid var(--color-grey2);
	border-radius: 0;
  transition-property: transform;
	transition-duration: 0.4s;
	width: 8.5vh;
	height: 8.5vh;
	padding: 0;
	margin: 0 auto;
	text-align: center;
	vertical-align: middle;
}
.drop.drop_ButtonBasic-Grey_Dark {
	background-color: #ffffff80;
	color: #ffffff80;
	border: 0.25vh solid #ffffff80;
	border-radius: 1.3vh;
	min-width: 4.5vh;
	min-height: 4.5vh;
	padding: 0;
}
.drop.drop_ButtonBasic-Grey_Light {
	background-color: #ffffff40;
	color: #ffffff80;
	border: 0.25vh solid #ffffff80;
	border-radius: 1.3vh;
	min-width: 4.5vh;
	min-height: 4.5vh;
	padding: 0;
}
</style>
