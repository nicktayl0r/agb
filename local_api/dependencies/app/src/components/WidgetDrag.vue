<template>
	<fieldset
    :id="id"
    v-show="visibleStore"
    :disabled="disabled"
    draggable="true"
    @dragstart="dragstart"
    @dragenter="dragenter"
    class="drag"
    :class="AddRemoveClasses"
    :style="{order: childIndex}">
		<span class="slot">
			<slot />
		</span>
	</fieldset>
</template>

<script lang="ts">
// I chose fieldset over span because it has a disabled attribute
import { polyfill } from "mobile-drag-drop";
import { scrollBehaviourDragImageTranslateOverride } from "mobile-drag-drop/scroll-behaviour";
import { isMobileSafari } from "@/helpers/browserHelpers";
import { logError, logWarning } from "@/helpers/debugHelpers";
import { removeDragWidget, addDragWidget } from "@/components/WidgetDrop.vue";

polyfill({
	// use this to make use of the scroll behaviour
	dragImageTranslateOverride: scrollBehaviourDragImageTranslateOverride,
	forceApply: true,
});

//@ts-ignore
window.doMobileSafariDragScrollWorkaround = false;

import { Maybe } from "true-myth";
import { Component, Prop } from "vue-property-decorator";

import Widget from "./Widget";
import { canAccessTopWindow } from "@/helpers/browserHelpers";
import { logMessage } from "@/helpers/debugHelpers";
import WidgetEnabled from "@/mixins/WidgetEnabled";
import WidgetSharedData from "@/mixins/WidgetSharedData";
import { readWidgetEntryVal, dispatchUpdateWidget } from "@/store/modules/userData";
import { getStore } from "@/helpers/storeHelpers";
import { getCurrentPageID } from "@/helpers/userHelpers";

@Component({ mixins: [WidgetEnabled, WidgetSharedData] })
export default class WidgetDrag extends Widget {
	@Prop({ default: "" })
	dropId: string;

	@Prop({ default: 0 })
	childIndex: number;

  @Prop({ default: "" })
	groupId: string;

  @Prop({ default: "" })
  correctDropIds: string;
	
  prevDropId: string;

	mounted() {
		const setParentDOM = this.dropIDAndSetDOM;
		if (this.$props.sharedDataReadKey) {
			const sharedValueOveride = this.propSharedDataOverride(
				"dropId",
				this.$props.sharedDataReadKey
			);
			this.updateWidget("dropId", sharedValueOveride);
			if (this.$props.sharedDataWriteKey) {
				this.updateSharedDataVal(
					this.$props.sharedDataWriteKey,
					sharedValueOveride
				);
			}
		}

		// prevent child images from doing their default drag behavior
		const slot = this.$el.querySelector(".slot");
		if (slot) {
			const childIMGs = slot.getElementsByTagName("IMG");
			for (let i = 0; i < childIMGs.length; i++) {
				childIMGs[i].setAttribute("draggable", "false");
			}
		}

		this.addMobileSafariScrollWorkaround();
	}

	// this doesn't need to do anything apparently
	safariTouchmove() {}

	beforeDestroy() {
		window.removeEventListener("touchmove", this.safariTouchmove);
	}

	updated() {
		const setParentDOM = this.dropIDAndSetDOM;
	}

	dragstart(ev: DragEvent) {
		if (!this.propStoreOverride("enabled") || !this.notGrapes) return;

		const target = ev.target;
		if (ev.dataTransfer) {
      //@ts-ignore
			ev.dataTransfer.setData("draggedWidgetID", target.id);
			logMessage("shared data key: ", this.$props.sharedDataWriteKey);
			if (this.$props.sharedDataWriteKey != undefined)
				logMessage("drag and send shared data key");
			ev.dataTransfer.setData(
				"draggedWidgetSharedDataWriteKey",
				this.$props.sharedDataWriteKey
			);
		}
	}

	// needed by mobile polyfill
	dragenter(ev: DragEvent) {
		ev.preventDefault();
	}

  get dropIDAndSetDOM() {
		let parentDropId;
		if (this.$props.sharedDataReadKey) {
			parentDropId = this.propSharedDataOverride(
				"dropId",
				this.$props.sharedDataReadKey
			);
		} else parentDropId = this.propStoreOverride("dropId");

		// remove prev drop dragId in store for DropsHaveDragsComparison
		if (this.prevDropId && this.prevDropId !== parentDropId) {
			removeDragWidget(this.prevDropId, this.id, this.pageID);
		}

		if (typeof parentDropId === "string" && parentDropId !== "") {
			if (parentDropId != this.prevDropId) {
				this.setParentDOM(parentDropId);
				// set drop dragId in store for DropsHaveDragsComparison
				addDragWidget(parentDropId, this.id, this.pageID);
			}
		} else {
			logWarning(
				"dropParent is not set on WidgetDrag",
				this.id,
				"has value",
				parentDropId
			);
		}
		this.prevDropId = parentDropId.toString();
		return parentDropId;
	}

	setParentDOM(parentDropId: string) {
		const dropEl = document.getElementById(parentDropId);
		if (!dropEl) return;
		const parentElement = dropEl.getElementsByClassName("dragSlot")[0];
		if (parentElement) {
			parentElement.appendChild(this.$el);
		} else {
			console.warn(
				"couldn't find element for dropParentID",
				parentDropId,
				"on WidgetDrag",
				this.id
			);
		}
	}

	addMobileSafariScrollWorkaround() {
		logMessage(
			"doMobileSafariDragScrollWorkaround",
      //@ts-ignore
			window.doMobileSafariDragScrollWorkaround,
			"isMobileSafari",
			isMobileSafari()
		);

      //@ts-ignore
		if (window.doMobileSafariDragScrollWorkaround || isMobileSafari()) {
			try {
				// workaround for iOS11.3/iOS12 touchmove behaviour (https://github.com/timruffles/mobile-drag-drop/issues/124)

				// Test via a getter in the options object to see if the passive property is accessed
				let supportsPassive = false;
				try {
					const opts = Object.defineProperty({}, "passive", {
						get() {
							supportsPassive = true;
						},
					});
					window.addEventListener("testPassive", () => {}, opts);
					window.removeEventListener("testPassive", () => {}, opts);
				} catch (e) {}

				const canAccessTop = canAccessTopWindow();

				logMessage(
					"supportsPasssive",
					supportsPassive,
					"canAccessTop",
					canAccessTop
				);

				if (supportsPassive) {
					window.addEventListener("touchmove", this.safariTouchmove, {
						passive: false,
					});
					// if the case is running in an iFrame, let's apply the iOS workaround to the top window as well
					if (canAccessTop) {
						if (window != window.top) {
							window.top.addEventListener("touchmove", this.safariTouchmove, {
								passive: false,
							});
						}
					}
				} else {
					window.addEventListener("touchmove", () => {});
					if (canAccessTop) {
						if (window != window.top) {
							window.top.addEventListener("touchmove", this.safariTouchmove);
						}
					}
				}
			} catch (err) {
				logError(err);
			}
		}
	}
}
</script>

<style>
/* basic drag "button" styles. (normal aka "diamond", ) */
.drag {
	display: inline-block;
}
.drag:disabled {
	cursor: not-allowed;
	pointer-events: none;
}
.drag:enabled:hover {
	cursor: grab;
}
.drag:enabled:active {
	cursor: grabbing;
}

.buttonDrag {
	background: transparent url(../assets/svg/buttonDrag_Normal.svg) no-repeat
		center;
	width: 8.5vh;
	height: 8.5vh;
	border: none;
	cursor: default;
	text-align: center;
	padding: 0;
	margin: 0 auto;
}
.buttonDrag:enabled:hover {
	background: transparent url(../assets/svg/buttonDrag_HoverDepressed.svg)
		no-repeat center;
}
.buttonDrag:disabled,
.buttonDrag_White:disabled {
	background: transparent url(../assets/svg/buttonDrag_Disabled.svg) no-repeat
		center;
	cursor: not-allowed;
}
.buttonDrag .slot {
	/* this is to vertical align slot contents */
	display: flex;
	justify-content: center;
	flex-direction: column;
	height: 100%;
}

/* .buttonDrag_White - START */
.buttonDrag_White {
	background: transparent url(../assets/svg/buttonDrag_White_Normal.svg)
		no-repeat center;
	width: 8.5vh;
	height: 8.5vh;
	border: none;
	cursor: default;
	text-align: center;
	padding: 0;
	margin: 0 auto;
}
.buttonDrag_White:enabled:hover {
	background: transparent url(../assets/svg/buttonDrag_White_HoverDepressed.svg)
		no-repeat center;
}
/* .buttonDrag_White - END */

.drag.buttonBasic-Grey {
	display: flex;
	justify-content: center;
	align-items: center;
  transition-property: transform;
	transition-duration: 0.4s;
	background: linear-gradient(var(--color-grey5), var(--color-grey6));
	border: 0.25vh solid var(--color-grey6);
	padding: 1.2vh;
	color: var(--color-white);
	margin: 0px;
	font-size: inherit;
	min-width: 4.5vh;
	min-height: 4.5vh;
	line-height: 1.2em;
	border-radius: 1.3vh;
}
.drag.buttonBasic-Grey:enabled:active {
	background: linear-gradient(var(--color-grey5), var(--color-grey6));
}
.drag.buttonBasic-Grey:enabled:hover {
	box-shadow: 0 0 0 0.45vh var(--color-blue3);
}
.drag.buttonBasic-Grey:disabled {
	background: linear-gradient(var(--color-grey4), var(--color-grey5));
	border: 0.25vh solid var(--color-grey5);
	opacity: 0.5;
}
.buttonBlinkBlue.drag:enabled {
	animation-name: blinkBlue;
	animation-duration: 1s;
	animation-iteration-count: infinite;
	animation-timing-function: ease-out;
	animation-direction: alternate;
}

.buttonBlinkBlue.drag:enabled:hover {
	animation-name: none;
}
</style>