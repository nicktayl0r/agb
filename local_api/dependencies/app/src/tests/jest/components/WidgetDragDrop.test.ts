import { getLocalVue, getCleanStore } from "@/tests/jest/jestSetup";
const localVue = getLocalVue();
let store = getCleanStore();

import { mount } from "@vue/test-utils";
import { WidgetBaseTests } from "@/tests/jest/components/WidgetBaseTests";
import { MixinWidgetEnabledTests } from "@/tests/jest/mixins/MixinWidgetEnabledTests";

import WidgetDrag from "@/components/WidgetDrag.vue";
import WidgetDrop from "@/components/WidgetDrop.vue";
import { MockDragEvent } from "@/tests/jest/__mock__/Event";
import {
  readWidgetEntryVal,
  dispatchUpdateWidget
} from "@/store/modules/userData";

let elem: any;


describe("WidgetDragDrop", () => {
  beforeEach(() => {
    store = getCleanStore();
    elem = document.createElement('div');
    if (document.body) {
      document.body.appendChild(elem);
    }
  });

  afterEach(() => {

  });

  WidgetBaseTests(WidgetDrag);
  MixinWidgetEnabledTests(WidgetDrag);

  WidgetBaseTests(WidgetDrop);
  MixinWidgetEnabledTests(WidgetDrop);

  it("matches html snapshot", async () => {
    const wrapperDrop = mount(WidgetDrop, {
      propsData: {
        id: "testDrop"
      },
      localVue,
      store,
      attachTo: elem
    });
    const wrapperDrag = mount(WidgetDrag, {
      propsData: {
        id: "testDrag",
        dropId: "testDrop"
      },
      localVue,
      store
    });
    await wrapperDrag.vm.$nextTick();

    expect(wrapperDrop.find('.dragSlot').find("#testDrag")).not.toBeNull();
    expect((wrapperDrop.find('.dragSlot').find("#testDrag") as any).id).toBe((wrapperDrag.vm.$el.parentElement as any).selector);
    expect(wrapperDrop.element).toMatchSnapshot();
    wrapperDrop.destroy();
  });
});

//DONE test that when dropped, a drag sets the right dropID in the store. Also that the dragID is in the drop.dragID, and also removed from the first drop
//DONE test that if disabled, a drop doesn't accept a drag
//e2e - test that drags dropped over a drop.maxChildren are swapped out (track store ids)
//DONE test that if a drop.dragID is a string, that it gets converted to an array with that string

describe("WidgetDragDrop interaction", () => {
  beforeEach(() => {
    store = getCleanStore();
    elem = [document.createElement('div'), document.createElement('div'), document.createElement('div'), document.createElement('div')];
    if (document.body) {
      elem.forEach((element: any) => {
        document.body.appendChild(element);
      });
    }
  });

  it("drag.dropId is correct", async () => {
    const dropStart = mount(WidgetDrop, {
      propsData: {
        id: "dropStart",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[0]
    });
    const dropEnd = mount(WidgetDrop, {
      propsData: {
        id: "dropEnd",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[1]
    });
    const drag1 = mount(WidgetDrag, {
      propsData: {
        id: "drag1",
        dropId: "dropStart", //attach to wrapperDropStart to begin
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[2]
    });
    const dropEvent = MockDragEvent("drop");
    //@ts-ignore
    dropEvent.dataTransfer.setData("draggedWidgetID", "drag1");
    dropEnd.vm.$el.dispatchEvent(dropEvent);

    await drag1.vm.$nextTick();
    await dropStart.vm.$nextTick();
    await dropEnd.vm.$nextTick();

    const startDrag1dropId = readWidgetEntryVal(store)(
      "testPage",
      "drag1",
      "dropId"
    );
    expect(startDrag1dropId.unsafelyUnwrap()).toBe("dropEnd");
    expect(drag1.vm.$el.parentElement!.parentElement!.id).toBe("dropEnd");

    drag1.destroy();
    dropStart.destroy();
    dropEnd.destroy();
  });

  it("drop.dragId is an array and correct", async () => {
    const dropStart = mount(WidgetDrop, {
      propsData: {
        id: "dropStart",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[0]
    });
    const dropEnd = mount(WidgetDrop, {
      propsData: {
        id: "dropEnd",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[1]
    });
    const drag1 = mount(WidgetDrag, {
      propsData: {
        id: "drag1",
        dropId: "dropStart", //attach to wrapperDropStart to begin
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[2]
    });
    const dropEvent = MockDragEvent("drop");
    //@ts-ignore
    dropEvent.dataTransfer.setData("draggedWidgetID", "drag1");
    dropEnd.vm.$el.dispatchEvent(dropEvent);

    await drag1.vm.$nextTick();
    await dropStart.vm.$nextTick();
    await dropEnd.vm.$nextTick();

    const startDrag1dropId = readWidgetEntryVal(store)(
      "testPage",
      "dropEnd",
      "dragId"
    );
    expect(startDrag1dropId.unsafelyUnwrap()).toStrictEqual(["drag1"]);

    drag1.destroy();
    dropStart.destroy();
    dropEnd.destroy();
  });

  it("drop.dragId is removed from array", async () => {
    const dropStart = mount(WidgetDrop, {
      propsData: {
        id: "dropStart",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[0]
    });
    const drop2 = mount(WidgetDrop, {
      propsData: {
        id: "drop2",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[1]
    });
    const drag1 = mount(WidgetDrag, {
      propsData: {
        id: "drag1",
        dropId: "dropStart", //attach to wrapperDropStart to begin
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[2]
    });
    const dropEvent = MockDragEvent("drop");
    //@ts-ignore
    dropEvent.dataTransfer.setData("draggedWidgetID", "drag1");
    //first drop on drop2
    drop2.vm.$el.dispatchEvent(dropEvent);

    await drag1.vm.$nextTick();
    await dropStart.vm.$nextTick();
    await drop2.vm.$nextTick();

    const dropStartDrag1dropId = readWidgetEntryVal(store)(
      "testPage",
      "dropStart",
      "dragId"
    );
    //confirm that drag1 moved to drop2

    expect(dropStartDrag1dropId.unsafelyUnwrap()).toStrictEqual([]);

    drag1.destroy();
    dropStart.destroy();
    drop2.destroy();
  });

  it("disabled drop doesn't accept drag", async () => {
    const dropStart = mount(WidgetDrop, {
      propsData: {
        id: "dropStart",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[0]
    });
    const drop2 = mount(WidgetDrop, {
      propsData: {
        id: "drop2",
        pageID: "testPage",
        enabled: false
      },
      localVue,
      store,
      attachTo: elem[1]
    });
    const drag1 = mount(WidgetDrag, {
      propsData: {
        id: "drag1",
        dropId: "dropStart", //attach to wrapperDropStart to begin
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[2]
    });
    const dropEvent = MockDragEvent("drop");
    //@ts-ignore
    dropEvent.dataTransfer.setData("draggedWidgetID", "drag1");
    //first drop on drop2
    drop2.vm.$el.dispatchEvent(dropEvent);

    await drag1.vm.$nextTick();
    await dropStart.vm.$nextTick();
    await drop2.vm.$nextTick();

    const dropStartDrag1dropId = readWidgetEntryVal(store)(
      "testPage",
      "drop2",
      "dragId"
    );
    //confirm that drag1 is not stored in drop2
    expect(dropStartDrag1dropId.isNothing()).toBe(true);

    drag1.destroy();
    dropStart.destroy();
    drop2.destroy();
  });

  it("backwardsCompat: string dragId is perserved in array", async () => {
    dispatchUpdateWidget(store, {
      pageID: "testPage",
      widgetID: "dropStart",
      key: "dragId",
      value: "testId"
    });
    const dropStart = mount(WidgetDrop, {
      propsData: {
        id: "dropStart",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[0]
    });
    const drop2 = mount(WidgetDrop, {
      propsData: {
        id: "drop2",
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[1]
    });
    const drag1 = mount(WidgetDrag, {
      propsData: {
        id: "drag1",
        dropId: "dropStart", //attach to dropStart to begin
        pageID: "testPage"
      },
      localVue,
      store,
      attachTo: elem[2]
    });

    await drag1.vm.$nextTick();
    await dropStart.vm.$nextTick();
    await drop2.vm.$nextTick();

    const dropStartDrag1dropId = readWidgetEntryVal(store)(
      "testPage",
      "dropStart",
      "dragId"
    );

    //confirm that drag2 is now in dropStart since it was swapped out.
    expect(dropStartDrag1dropId.unsafelyUnwrap()).toStrictEqual([
      "testId",
      "drag1"
    ]);

    drag1.destroy();
    dropStart.destroy();
    drop2.destroy();
  });
});
