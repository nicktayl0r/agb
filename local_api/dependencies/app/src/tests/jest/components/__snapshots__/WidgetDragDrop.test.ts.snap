// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetDragDrop matches html snapshot 1`] = `
<fieldset
  class="drop"
  id="testDrop"
  maxchildren="1"
  overflowstyle="reset-child"
>
  <span
    class="contentSlot noSelect"
  />
   
  <span
    class="dragSlot"
  >
    <fieldset
      class="drag"
      draggable="true"
      id="testDrag"
      style="order: 0;"
    >
      <span
        class="slot"
      />
    </fieldset>
  </span>
</fieldset>
`;
